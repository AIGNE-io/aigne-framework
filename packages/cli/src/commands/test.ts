import { spawnSync } from "node:child_process";
import { isAbsolute, resolve } from "node:path";
import { Command } from "commander";

export function createTestCommand(): Command {
  return new Command("test")
    .description("Run tests in the specified agents directory")
    .option(
      "--url, --path <path_or_url>",
      "Path to the agents directory or URL to aigne project",
      ".",
    )
    .action(async (options: { path: string }) => {
      const { path } = options;
      const absolutePath = isAbsolute(path) ? path : resolve(process.cwd(), path);

      spawnSync("node", ["--test"], { cwd: absolutePath, stdio: "inherit" });
    })
    .showHelpAfterError(true)
    .showSuggestionAfterError(true);
}
