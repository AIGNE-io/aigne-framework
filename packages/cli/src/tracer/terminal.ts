import { inspect } from "node:util";
import {
  type Agent,
  ChatModel,
  type ChatModelOutput,
  type Context,
  type Message,
} from "@aigne/core";
import type { ContextUsage } from "@aigne/core/execution-engine/usage";
import {
  type DefaultRenderer,
  Listr,
  ListrDefaultRendererLogLevels,
  type ListrRenderer,
  type ListrTaskWrapper,
  Spinner,
  figures,
} from "@aigne/listr2";
import chalk from "chalk";
import { z } from "zod";
import { parseDuration } from "../utils/time.js";

const DEBUG_DEPTH = z.number().int().safeParse(Number(process.env.DEBUG_DEPTH)).data;

export class TerminalTracer {
  constructor(public readonly context: Context) {}

  private spinner = new Spinner();

  private tasks: { [callId: string]: Task } = {};

  async run(agent: Agent, input: Message) {
    try {
      this.spinner.start();

      const context = this.context.newContext({ reset: true });

      const listr = this.newListr();

      context.on(
        "agentStarted",
        async ({ contextId, parentContextId, agent, input, timestamp }) => {
          const task: Task = {
            ...Promise.withResolvers(),
            listr: Promise.withResolvers(),
            startTime: timestamp,
          };
          this.tasks[contextId] = task;

          const listrTask: Parameters<typeof listr.add>[0] = {
            title: this.formatTaskTitle(agent),
            task: (ctx, taskWrapper) => {
              const subtask = taskWrapper.newListr([{ task: () => task.promise }]);
              task.listr.resolve({ subtask, taskWrapper, ctx });
              return subtask;
            },
            rendererOptions: {
              persistentOutput: true,
              outputBar: Number.POSITIVE_INFINITY,
              bottomBar: Number.POSITIVE_INFINITY,
            },
          };

          const parentTask = parentContextId ? this.tasks[parentContextId] : undefined;
          if (parentTask) {
            parentTask.listr.promise.then(({ subtask }) => {
              subtask.add(listrTask);
            });
          } else {
            listr.add(listrTask);
          }

          const { taskWrapper } = await task.listr.promise;

          taskWrapper.output = this.formatAgentStartedOutput(agent, input);
        },
      );

      context.on(
        "agentSucceed",
        async ({ agent, contextId, parentContextId, output, timestamp }) => {
          const task = this.tasks[contextId];
          if (!task) return;

          task.endTime = timestamp;

          const { taskWrapper, ctx } = await task.listr.promise;

          if (agent instanceof ChatModel) {
            const { usage } = output as ChatModelOutput;
            task.usage = usage;
          }

          taskWrapper.title = this.formatTaskTitle(agent, { task, usage: true, time: true });
          taskWrapper.output = this.formatAgentSucceedOutput(agent, output);

          if (!parentContextId || !this.tasks[parentContextId]) {
            Object.assign(ctx, output);
          }

          task.resolve();
        },
      );

      context.on("agentFailed", async ({ agent, contextId, error, timestamp }) => {
        const task = this.tasks[contextId];
        if (!task) return;

        task.endTime = timestamp;

        const { taskWrapper } = await task.listr.promise;
        taskWrapper.title = this.formatTaskTitle(agent, { task, usage: true, time: true });
        taskWrapper.output = this.formatAgentFailedOutput(agent, error);

        task.reject(error);
      });

      const [result] = await Promise.all([
        listr.waitTaskAndRun(),
        context.call(agent, input).finally(() => {
          listr.resolveWaitingTask();
        }),
      ]);

      return { result, context };
    } finally {
      this.spinner.stop();
    }
  }

  protected newListr() {
    return new MyListr([], {
      concurrent: true,
      rendererOptions: {
        collapseSubtasks: false,
        writeBottomBarDirectly: true,
        icon: {
          [ListrDefaultRendererLogLevels.PENDING]: () => this.spinner.fetch(),
          [ListrDefaultRendererLogLevels.OUTPUT_WITH_BOTTOMBAR]: "",
        },
      },
    });
  }

  formatAgentStartedOutput(agent: Agent, data: Message) {
    return `\
${chalk.grey(figures.pointer)} call agent ${agent.name} started with input:
${this.formatMessage(data)}`;
  }

  formatAgentSucceedOutput(agent: Agent, data: Message) {
    return `\
${chalk.grey(figures.tick)} call agent ${agent.name} succeed with output:
${this.formatMessage(data)}`;
  }

  formatAgentFailedOutput(agent: Agent, data: Error) {
    return `\
${chalk.grey(figures.cross)} call agent ${agent.name} failed with error:
${this.formatMessage(data)}`;
  }

  formatMessage(data: unknown) {
    return inspect(data, { colors: true, depth: DEBUG_DEPTH });
  }

  formatTokenUsage(usage: Partial<ContextUsage>) {
    const items = [
      [chalk.yellow(usage.promptTokens), chalk.grey("prompt tokens")],
      [chalk.cyan(usage.completionTokens), chalk.grey("completion tokens")],
      usage.agentCalls ? [chalk.magenta(usage.agentCalls), chalk.grey("agent calls")] : undefined,
    ];

    const content = items
      .filter((i) => !!i)
      .map((i) => i.join(" "))
      .join(chalk.grey(", "));

    return `${chalk.grey("(Usage: ")}${content}${chalk.grey(")")}`;
  }

  formatTimeUsage(startTime: number, endTime: number) {
    const duration = endTime - startTime;
    return chalk.grey(`[${parseDuration(duration)}]`);
  }

  formatTaskTitle(
    agent: Agent,
    { task, usage, time }: { task?: Task; usage?: boolean; time?: boolean } = {},
  ) {
    let title = `call agent ${agent.name}`;

    if (usage && task?.usage) title += ` ${this.formatTokenUsage(task.usage)}`;
    if (time && task?.startTime && task.endTime)
      title += ` ${this.formatTimeUsage(task.startTime, task.endTime)}`;

    return title;
  }
}

type Task = ReturnType<typeof Promise.withResolvers<void>> & {
  listr: ReturnType<
    typeof Promise.withResolvers<{
      ctx: object;
      subtask: Listr;
      taskWrapper: ListrTaskWrapper<unknown, typeof DefaultRenderer, typeof ListrRenderer>;
    }>
  >;
  startTime?: number;
  endTime?: number;
  usage?: Partial<ContextUsage>;
};

class MyListr extends Listr {
  private taskPromise = Promise.withResolvers();
  private isTaskPromiseResolved = false;

  resolveWaitingTask() {
    if (!this.isTaskPromiseResolved) {
      this.taskPromise.resolve();
      this.isTaskPromiseResolved = true;
    }
  }

  override add(...args: Parameters<Listr["add"]>): ReturnType<Listr["add"]> {
    const result = super.add(...args);
    this.resolveWaitingTask();
    return result;
  }

  async waitTaskAndRun(ctx?: unknown) {
    if (!this.tasks.length) await this.taskPromise.promise;

    if (!this.tasks.length) return ctx;

    return super.run(ctx);
  }
}
