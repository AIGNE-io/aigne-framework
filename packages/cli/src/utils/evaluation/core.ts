import { Listr } from "@aigne/listr2";
import { ConsoleReporter } from "./reporter.js";
import type {
  Dataset,
  DatasetItem,
  Evaluation,
  EvaluationSummary,
  Evaluator,
  Report,
  Reporter,
  Runner,
  RunOptions,
  RunResult,
} from "./type.ts";

function aggregateSummary(
  results: (RunResult & { evaluations: Evaluation[]; error?: any })[],
): EvaluationSummary {
  const total = results.length;
  const scores = results.flatMap((r) => r.evaluations.map((e) => e.score));
  const successRate = Number(
    (scores.length > 0 ? scores.reduce((a, b) => a + b, 0) / scores.length : 0).toFixed(2),
  );

  const latencies = results.map((r) => r.latency || 0);
  const totalTokens = results.reduce(
    (a, r) => a + (r.usage?.inputTokens || 0) + (r.usage?.outputTokens || 0),
    0,
  );

  const errors = results.filter((r) => r.error).length;

  return {
    total,
    successRate,
    avgLatency: latencies.reduce((a, b) => a + b, 0) / (latencies.length || 1),
    maxLatency: Math.max(...latencies, 0),
    minLatency: Math.min(...latencies, 0),
    totalTokens,
    errorCount: errors,
    scoreDistribution: {
      min: Math.min(...scores, 0),
      max: Math.max(...scores, 0),
      mean: successRate,
      median: scores.length ? scores.sort((a, b) => a - b)[Math.floor(scores.length / 2)] : 0,
      variance:
        scores.length > 1
          ? scores.reduce((a, b) => a + (b - successRate) ** 2, 0) / scores.length
          : 0,
    },
  };
}

type EvaluationPipelineContext = {
  items: DatasetItem[];
  results: (RunResult & { evaluations: Evaluation[] })[];
};

export async function runEvaluationPipeline(params: {
  dataset: Dataset;
  runner: Runner;
  evaluators: Evaluator[];
  reporters?: Reporter[];
  options?: RunOptions;
}) {
  const { dataset, runner, evaluators, reporters = [new ConsoleReporter()], options } = params;

  const results: EvaluationPipelineContext["results"] = [];

  const task1 = new Listr<{ items: DatasetItem[] }>([
    {
      title: "Load dataset",
      task: async (ctx, _task) => {
        ctx.items = await dataset.loadWithOptions();
      },
    },
  ]);

  const { items } = await task1.run();

  const task2 = new Listr<EvaluationPipelineContext>(
    items.map((item) => {
      const input = JSON.stringify(item.input);
      return {
        title: `Run evaluations with input: ${input.length > 100 ? `${input.slice(0, 100)}...` : input}...`,
        task: async (ctx, task) => {
          task.output = `Start running agent with input: ${JSON.stringify(item.input, null, 2)}`;

          const runnerResults = runner.run([item], options);

          for await (const result of runnerResults) {
            task.output = `Start running evaluation with: ${JSON.stringify(
              {
                input: result.input,
                output: result.output,
                expected: result.expected,
              },
              null,
              2,
            )}`;

            const evaluations: Evaluation[] = [];
            for (const evaluator of evaluators) {
              const evals = await evaluator.evaluate(result);
              evaluations.push(...evals);
            }

            results.push({ ...result, evaluations });

            task.output = `Finish running evaluation`;
          }

          ctx.results = results;
        },
      };
    }),
    {
      concurrent: options?.concurrency ? Math.min(items.length, options?.concurrency) : false,
      exitOnError: true,
      rendererOptions: {
        collapseSubtasks: false,
      },
    },
  );

  await task2.run();

  const summary: EvaluationSummary = aggregateSummary(results);
  const report: Report = { dataset: dataset.name, results, summary };

  for (const reporter of reporters) {
    await reporter.report(report);
  }
}
