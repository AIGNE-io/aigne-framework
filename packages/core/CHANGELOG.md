## 1.1.0-beta.17 (2025-3-18)

- chore: add support for esm module


## [1.2.0](https://github.com/AIGNE-io/aigne-framework/compare/core-next-v1.1.0...core-next-v1.2.0) (2025-03-18)


### Features

* add Agent FunctionAgent AIAgent MCPAgent and ExecutionEngine ([4d2a5a1](https://github.com/AIGNE-io/aigne-framework/commit/4d2a5a1b3366b8f935f50a0937c2da6e49073348))
* add OrchestratorAgent in agent library ([25a5e9e](https://github.com/AIGNE-io/aigne-framework/commit/25a5e9e6c60d747c8bf484ac884b31dc02c14757))
* add sequential and parallel helper function ([a295697](https://github.com/AIGNE-io/aigne-framework/commit/a295697b5694754e02954fc5c7f382a3b219a3ab))
* add support for MCP resources ([1ded2fb](https://github.com/AIGNE-io/aigne-framework/commit/1ded2fbf222fa8984e75df0852ff384524f73b04))
* **prompt-builder:** support chat history in PromptBuilder ([6ca05f2](https://github.com/AIGNE-io/aigne-framework/commit/6ca05f28eddb683a4f1e228865f8bbf8a8e190f1))
* support run puppeteer example chat loop in terminal ([85ce7f8](https://github.com/AIGNE-io/aigne-framework/commit/85ce7f8de8b443c86e50815dd7bcab99f869c4ce))
* use PromptBuilder instead of string instructions ([e4cb2cb](https://github.com/AIGNE-io/aigne-framework/commit/e4cb2cb6baf4f9bcef390567a4a174e9246c29a3))


### Bug Fixes

* **AIAgent:** should pass both arguments (model generated) and input (user input) to tool ([c49d64e](https://github.com/AIGNE-io/aigne-framework/commit/c49d64ee35f7efd83b0f82f43205bb1c40f999e8))
* **MCP:** catch list resource error treat as empty list ([1885fab](https://github.com/AIGNE-io/aigne-framework/commit/1885fab3585e0dd1467b127e5b47cd0b98282bab))
* use text resource from MCP correctly ([8b9eba8](https://github.com/AIGNE-io/aigne-framework/commit/8b9eba83352ec096a2a5d4f410d4c4bde7420bce))

## 1.1.0-beta.16 (2025-3-18)

- chore: add puppeteer in linux need docker_container

## 1.1.0-beta.15 (2025-3-18)

- chore: make coverage report as text to terminal
- chore: update contributing docs

## 1.1.0-beta.14 (2025-3-18)

- chore(example): add code-execution example

## 1.1.0-beta.13 (2025-3-18)

- feat: add OrchestratorAgent in agent library

## 1.1.0-beta.12 (2025-3-14)

- chore(example): add concurrency reflection handoff workflow examples

## 1.1.0-beta.11 (2025-3-14)

- feat(core): add sequential and parallel helper function
- chore(examples): add workflow-sequential example

## 1.1.0-beta.10 (2025-3-13)

- chore: ensure required environment variables have values

## 1.1.0-beta.9 (2025-3-13)

- fix(MCP): catch list resource error treat as empty list

## 1.1.0-beta.8 (2025-3-13)

- fix(AIAgent): should pass both arguments (model generated) and input (user input) to tool
- chore(examples): add workflow-router example
- chore(examples): rename examples puppeteer-mcp-server and sqlite-mcp-server to mcp-server-puppeteer and mcp-server-sqlite

## 1.1.0-beta.7 (2025-3-13)

- chore: rename @aigne/core to @aigne/core-next

## 1.1.0-beta.6 (2025-3-13)

- chore(examples): default enable mcp debug message for examples

## 1.1.0-beta.5 (2025-3-13)

- feat: support chat history in PromptBuilder
- feat: add `prompts` for MCPAgent to consume prompts from MCP server
- chore: add sqlite-mcp-server example
- test: add more unit test cases

## 1.1.0-beta.4 (2025-3-12)

- feat: support run puppeteer example chat loop in terminal

## 1.1.0-beta.3 (2025-3-11)

- chore: set module type for core package

## 1.1.0-beta.2 (2025-3-11)

- feat: use PromptBuilder instead of string instructions
- refactor: use tools instead of skills
- chore(examples): add puppeteer-mcp-server example

## 1.1.0-beta.1 (2025-3-11)

- feat: add Agent FunctionAgent AIAgent MCPAgent and ExecutionEngine
