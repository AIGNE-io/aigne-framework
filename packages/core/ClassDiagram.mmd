classDiagram
    class PromptBuilderBuildOptions {
        +Context context
        +Agent agent
        +object input
        +ChatModel model
    }

    class Prompt {
        +List~object~ messages
        +List~Agent~ tools
        +object toolChoice
        +object responseFormat
    }

    PromptBuilder ..> PromptBuilderBuildOptions: dependency
    PromptBuilder ..> Prompt: dependency
    class PromptBuilder {
        +build(PromptBuilderBuildOptions options) Prompt
    }

    class Model {
        +call(Prompt input) object*
    }

    ChatModel --|> Model: inheritance
    class ChatModel {
    }

    ImageModel --|> Model: inheritance
    class ImageModel {
    }

    Agent --|> EventEmitter: inheritance
    class Agent {
        +string name
        +string description
        +object inputSchema
        +object outputSchema
        +List~string~ inputTopic
        +List~string~ nextTopic
        +List~Agent~ tools
        +Map~string, Agent~ skills

        +call(string input, Context context) object
        +call(object input, Context context) object

        +process(object input, Context context) object*
        -verifyInput() void
        -verifyOutput() void
    }

    AIAgent --|> Agent: inheritance
    AIAgent *.. PromptBuilder: composition
    class AIAgent {
        +ChatModel model
        +string instructions
        +string outputKey
        +object toolChoice
        +PromptBuilder promptBuilder
    }

    ImageAgent --|> Agent: inheritance
    ImageAgent *.. PromptBuilder: composition
    class ImageAgent {
        +ImageModel model
        +string instructions
        +PromptBuilder promptBuilder
    }

    FunctionAgent --|> Agent: inheritance
    class FunctionAgent {
        +Function fn
    }

    RPCAgent --|> Agent: inheritance
    class RPCAgent {
        +string url
    }

    MCPAgent --|> Agent: inheritance
    MCPAgent *.. MCPClient: composition
    class MCPAgent {
        +MCPClient client
    }

    class MCPClient {
    }

    class Message {
        +object output
    }

    MessageQueue ..> Message: dependency
    class MessageQueue {
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    class History {
        +string id
        +string agentId
        +object input
        +object output
    }

    Context *.. ChatModel: composition
    Context *.. ImageModel: composition
    Context --|> MessageQueue: inheritance
    Context ..> History: dependency
    class Context {
        +ChatModel model
        +ImageModel imageModel
        +List~Agent~ tools

        +getHistories(string agentId, int limit) List~History~
        +addHistory(History history) void
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    class EventEmitter {
        +on(): void
        +emit(): void
    }

    UserAgent --|> Agent: inheritance
    class UserAgent {
    }

    class ExecutionEngineRunOptions {
        +boolean concurrency
    }

    ExecutionEngine --|> Context: inheritance
    ExecutionEngine --|> EventEmitter: inheritance
    ExecutionEngine ..> UserAgent: dependency
    ExecutionEngine ..> ExecutionEngineRunOptions: dependency
    class ExecutionEngine {
        +run(Agent agent) UserAgent
        +run(string input) object
        +run(object input) object
        +run(object input, Agent ...agents) object
        +run(object input, ExecutionEngineRunOptions options, Agent ...agents) object
    }
