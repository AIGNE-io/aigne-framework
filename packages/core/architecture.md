## Class Definitions

```mermaid
classDiagram
    class PromptBuilderBuildOptions {
        +Context context
        +Agent agent
        +object input
        +ChatModel model
    }

    class Prompt {
        +List~object~ messages
        +List~Agent~ tools
        +object toolChoice
        +object responseFormat
    }

    PromptBuilder ..> PromptBuilderBuildOptions: dependency
    PromptBuilder ..> Prompt: dependency
    class PromptBuilder {
        +build(PromptBuilderBuildOptions options) Prompt
    }

    class Model {
        +call(Prompt input) object*
    }

    ChatModel --|> Model: inheritance
    class ChatModel {
    }

    ImageModel --|> Model: inheritance
    class ImageModel {
    }

    Agent --|> EventEmitter: inheritance
    class Agent {
        +string name
        +string description
        +object inputSchema
        +object outputSchema
        +List~string~ inputTopic
        +List~string~ nextTopic
        +List~Agent~ tools

        +call(string input, Context context) object
        +call(object input, Context context) object

        +process(object input, Context context) object*
        -verifyInput() void
        -verifyOutput() void
    }

    AIAgent --|> Agent: inheritance
    AIAgent *.. PromptBuilder: composition
    class AIAgent {
        +ChatModel model
        +string instructions
        +string outputKey
        +object toolChoice
        +PromptBuilder promptBuilder
    }

    ImageAgent --|> Agent: inheritance
    ImageAgent *.. PromptBuilder: composition
    class ImageAgent {
        +ImageModel model
        +string instructions
        +PromptBuilder promptBuilder
    }

    FunctionAgent --|> Agent: inheritance
    class FunctionAgent {
        +Function fn
    }

    RPCAgent --|> Agent: inheritance
    class RPCAgent {
        +string url
    }

    MCPAgent --|> Agent: inheritance
    MCPAgent *.. MCPClient: composition
    class MCPAgent {
        +MCPClient client
    }

    class MCPClient {
    }

    class Message {
        +object output
    }

    MessageQueue ..> Message: dependency
    class MessageQueue {
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    class History {
        +string id
        +string agentId
        +object input
        +object output
    }

    Context *.. ChatModel: composition
    Context *.. ImageModel: composition
    Context --|> MessageQueue: inheritance
    Context ..> History: dependency
    class Context {
        +ChatModel model
        +ImageModel imageModel
        +List~Agent~ tools

        +getHistories(string agentId, int limit) List~History~
        +addHistory(History history) void
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    class EventEmitter {
        +on(): void
        +emit(): void
    }

    UserAgent --|> Agent: inheritance
    class UserAgent {
    }


    AIGNE --|> Context: inheritance
    AIGNE --|> EventEmitter: inheritance
    AIGNE ..> UserAgent: dependency
    class AIGNE {
        +call(Agent agent) UserAgent
        +call(Agent agent, string input) object
        +call(Agent agent, object input) object
    }
```
