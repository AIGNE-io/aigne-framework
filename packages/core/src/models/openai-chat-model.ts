import { nanoid } from "nanoid";
import OpenAI from "openai";
import type { ChatCompletionMessageParam, ChatCompletionTool } from "openai/resources";
import { parseJSON } from "../utils/json-schema.js";
import { isNonNullable } from "../utils/type-utils.js";
import {
  ChatModel,
  type ChatModelInput,
  type ChatModelInputMessage,
  type ChatModelInputTool,
  type ChatModelOptions,
  type ChatModelOutput,
  type Role,
} from "./chat-model.js";

const CHAT_MODEL_OPENAI_DEFAULT_MODEL = "gpt-4o-mini";

export interface OpenAIChatModelOptions {
  apiKey?: string;
  baseURL?: string;
  model?: string;
  modelOptions?: ChatModelOptions;
}

export class OpenAIChatModel extends ChatModel {
  constructor(public options?: OpenAIChatModelOptions) {
    super();
  }

  private _client?: OpenAI;

  private get client() {
    if (!this.options?.apiKey) throw new Error("Api Key is required for OpenAIChatModel");

    this._client ??= new OpenAI({
      baseURL: this.options.baseURL,
      apiKey: this.options.apiKey,
    });
    return this._client;
  }

  get modelOptions() {
    return this.options?.modelOptions;
  }

  async process(input: ChatModelInput): Promise<ChatModelOutput> {
    const res = await this.client.chat.completions.create({
      model: this.options?.model || CHAT_MODEL_OPENAI_DEFAULT_MODEL,
      temperature: input.modelOptions?.temperature ?? this.modelOptions?.temperature,
      top_p: input.modelOptions?.topP ?? this.modelOptions?.topP,
      frequency_penalty:
        input.modelOptions?.frequencyPenalty ?? this.modelOptions?.frequencyPenalty,
      presence_penalty: input.modelOptions?.presencePenalty ?? this.modelOptions?.presencePenalty,
      messages: await contentsFromInputMessages(input.messages),
      tools: toolsFromInputTools(input.tools),
      tool_choice: input.toolChoice,
      parallel_tool_calls: !input.tools?.length
        ? undefined
        : (input.modelOptions?.parallelToolCalls ?? this.modelOptions?.parallelToolCalls),
      response_format:
        input.responseFormat?.type === "json_schema"
          ? {
              type: "json_schema",
              json_schema: {
                ...input.responseFormat.jsonSchema,
                schema: jsonSchemaToOpenAIJsonSchema(input.responseFormat.jsonSchema.schema),
              },
            }
          : undefined,
      stream: true,
    });

    let text = "";
    const toolCalls: (NonNullable<ChatModelOutput["toolCalls"]>[number] & {
      args: string;
    })[] = [];

    for await (const chunk of res) {
      const choice = chunk.choices?.[0];

      if (choice?.delta.tool_calls?.length) {
        for (const call of choice.delta.tool_calls) {
          toolCalls[call.index] ??= {
            id: call.id || nanoid(),
            type: "function" as const,
            function: { name: "", arguments: {} },
            args: "",
          };
          const c = toolCalls[call.index];
          if (!c) throw new Error("Tool call not found");

          if (call.type) c.type = call.type;

          c.function.name = c.function.name + (call.function?.name || "");
          c.args = c.args.concat(call.function?.arguments || "");
        }
      }

      if (choice?.delta.content) text += choice.delta.content;
    }

    const result: ChatModelOutput = {};

    if (input.responseFormat?.type === "json_schema" && text) {
      result.json = parseJSON(text);
    } else {
      result.text = text;
    }

    if (toolCalls.length) {
      result.toolCalls = toolCalls.map(({ args, ...c }) => ({
        ...c,
        function: { ...c.function, arguments: parseJSON(args) },
      }));
    }

    return result;
  }
}

const ROLE_MAP: { [key in Role]: ChatCompletionMessageParam["role"] } = {
  system: "system",
  user: "user",
  agent: "assistant",
  tool: "tool",
} as const;

async function contentsFromInputMessages(
  messages: ChatModelInputMessage[],
): Promise<ChatCompletionMessageParam[]> {
  return messages.map(
    (i) =>
      ({
        role: ROLE_MAP[i.role],
        content:
          typeof i.content === "string"
            ? i.content
            : i.content
                ?.map((c) => {
                  if (c.type === "text") {
                    return { type: "text" as const, text: c.text };
                  }
                  if (c.type === "image_url") {
                    return {
                      type: "image_url" as const,
                      image_url: { url: c.url },
                    };
                  }
                })
                .filter(isNonNullable),
        tool_calls: i.toolCalls?.map((i) => ({
          ...i,
          function: {
            ...i.function,
            arguments: JSON.stringify(i.function.arguments),
          },
        })),
        tool_call_id: i.toolCallId,
        name: i.name,
      }) as ChatCompletionMessageParam,
  );
}

function toolsFromInputTools(tools?: ChatModelInputTool[]): ChatCompletionTool[] | undefined {
  return tools?.length
    ? tools.map((i) => ({
        type: "function",
        function: {
          name: i.function.name,
          description: i.function.description,
          parameters: i.function.parameters as Record<string, unknown>,
        },
      }))
    : undefined;
}

function jsonSchemaToOpenAIJsonSchema(schema: Record<string, unknown>): Record<string, unknown> {
  if (schema?.type === "object") {
    const { required, properties } = schema as {
      required?: string[];
      properties: Record<string, unknown>;
    };

    return {
      ...schema,
      properties: Object.fromEntries(
        Object.entries(properties || {}).map(([key, value]) => {
          const valueSchema = jsonSchemaToOpenAIJsonSchema(value as Record<string, unknown>);

          // NOTE: All fields must be required https://platform.openai.com/docs/guides/structured-outputs/all-fields-must-be-required
          return [
            key,
            required?.includes(key) ? valueSchema : { anyOf: [valueSchema, { type: ["null"] }] },
          ];
        }),
      ),
      required: Object.keys(properties || {}),
    };
  }

  if (schema?.type === "array") {
    const { items } = schema as { items: Record<string, unknown> };

    return {
      ...schema,
      items: jsonSchemaToOpenAIJsonSchema(items),
    };
  }

  return schema;
}
