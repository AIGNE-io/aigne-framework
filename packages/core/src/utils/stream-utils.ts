import type {
  AgentProcessAsyncGeneratorResult,
  AgentResponseChunk,
  AgentResponseStream,
  Message,
} from "../agents/agent.js";

export function objectToAgentResponseStream<T extends Message>(json: T): AgentResponseStream<T> {
  return new ReadableStream({
    start(controller) {
      controller.enqueue({ delta: { json } });
      controller.close();
    },
  });
}

function mergeResult<T extends Message>(output: T, chunk: AgentResponseChunk<T>) {
  if (chunk.delta.text) {
    for (const [key, text] of Object.entries(chunk.delta.text)) {
      const original = output[key] as string | undefined;
      const t = (original || "") + (text || "");
      if (t) Object.assign(output, { [key]: t });
    }
  }

  if (chunk.delta.json) {
    Object.assign(
      output,
      Object.fromEntries(Object.entries(chunk.delta.json).filter(([_, v]) => v !== undefined)),
    );
  }
}

export async function agentResponseStreamToObject<T extends Message>(
  stream: AgentResponseStream<T> | AgentProcessAsyncGeneratorResult<T>,
): Promise<T> {
  const json: T = {} as T;

  if (stream instanceof ReadableStream) {
    const reader = stream.getReader();
    for (;;) {
      const { value, done } = await reader.read();
      if (done) break;
      mergeResult(json, value);
    }
  } else {
    for (;;) {
      const chunk = await stream.next();
      if (chunk.value) {
        if (chunk.done) {
          Object.assign(json, chunk.value);
        } else {
          mergeResult(json, chunk.value);
        }
      }
      if (chunk.done) break;
    }
  }

  return json;
}

export function asyncGeneratorToReadableStream<T extends Message>(
  generator: AgentProcessAsyncGeneratorResult<T>,
): AgentResponseStream<T> {
  return new ReadableStream({
    async start(controller) {
      try {
        if (generator instanceof ReadableStream) {
          for await (const value of generator) {
            controller.enqueue(value);
          }
        } else {
          for (;;) {
            const chunk = await generator.next();
            if (chunk.value) {
              if (chunk.done) {
                controller.enqueue({ delta: { json: chunk.value } });
              } else {
                controller.enqueue(chunk.value);
              }
            }
            if (chunk.done) break;
          }
        }
      } catch (error) {
        controller.error(error);
      } finally {
        controller.close();
      }
    },
  });
}

export function onAgentResponseStreamEnd<T extends Message>(
  stream: AgentResponseStream<T>,
  callback: (result: T) => unknown,
) {
  return new ReadableStream({
    async start(controller) {
      try {
        const json: T = {} as T;

        const reader = stream.getReader();

        for (;;) {
          const { value, done } = await reader.read();
          if (done) break;

          controller.enqueue(value);

          mergeResult(json, value);
        }

        Object.assign(json, await callback(json));
        controller.enqueue({
          delta: { json },
        });
      } catch (error) {
        controller.error(error);
      } finally {
        controller.close();
      }
    },
  });
}

export function isAsyncGenerator<T extends AsyncGenerator>(
  value: AsyncGenerator | unknown,
): value is T {
  return typeof value === "object" && value !== null && Symbol.asyncIterator in value;
}
