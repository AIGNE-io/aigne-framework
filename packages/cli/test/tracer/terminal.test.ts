import { expect, spyOn, test } from "bun:test";
import { TerminalTracer } from "@aigne/cli/tracer/terminal.js";
import { AIAgent, ExecutionEngine, createMessage } from "@aigne/core";
import { OpenAIChatModel } from "@aigne/core/models/openai-chat-model.js";
import { arrayToAgentProcessAsyncGenerator } from "@aigne/core/utils/stream-utils.js";

test("TerminalTracer should work correctly", async () => {
  const model = new OpenAIChatModel({});

  const engine = new ExecutionEngine({ model });
  const context = engine.newContext();

  const testAgent = AIAgent.from({});

  spyOn(model, "process").mockReturnValue(
    Promise.resolve({ text: "hello, this is a test response message" }),
  );

  const userAgent = engine.call(testAgent);

  const tracer = new TerminalTracer(context);

  const { result } = await tracer.run(userAgent, createMessage("hello"));

  expect(result).toMatchSnapshot();
});

test("TerminalTracer should raise error correctly", async () => {
  const engine = new ExecutionEngine();
  const context = engine.newContext();

  const testAgent = AIAgent.from({});

  spyOn(testAgent, "process").mockReturnValueOnce(
    arrayToAgentProcessAsyncGenerator([new Error("test error")]),
  );

  const userAgent = engine.call(testAgent);

  const tracer = new TerminalTracer(context);

  const result = tracer.run(userAgent, createMessage("hello"));

  expect(result).rejects.toThrowError("test error");
});

test("TerminalTracer should render output message with markdown highlight", async () => {
  const model = new OpenAIChatModel({});

  const engine = new ExecutionEngine({ model });
  const context = engine.newContext();

  const testAgent = AIAgent.from({});

  spyOn(model, "process").mockReturnValue(
    Promise.resolve({ text: "## Hello\nI am from [**AIGNE**](https://www.aigne.io)" }),
  );

  const userAgent = engine.call(testAgent);

  const tracer = new TerminalTracer(context);

  const formatAIResponse = spyOn(tracer, "formatAIResponse");

  await tracer.run(userAgent, createMessage("hello"));

  expect(formatAIResponse.mock.results.at(-1)).toMatchSnapshot();
});
