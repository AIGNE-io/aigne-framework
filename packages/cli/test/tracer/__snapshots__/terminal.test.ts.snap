// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`TerminalTracer should work correctly 1`] = `
{
  "$message": "hello, this is a test response message",
}
`;

exports[`TerminalTracer should render output message with markdown highlight 1`] = `
{
  "type": "return",
  "value": 
"\x1B[32m\x1B[1m## Hello\x1B[22m\x1B[39m

\x1B[0mI am from \x1B[34m\x1B[1mAIGNE\x1B[22m (\x1B[34m\x1B[4mhttps://www.aigne.io\x1B[24m\x1B[39m\x1B[34m)\x1B[39m\x1B[0m"
,
}
`;
